### PR TITLE
Add configuration option to specify custom Jinja2 markers.

### PR DESCRIPTION
**Pull request by makkus**
If a Markdown file already contains any of the default Jinja2 block or
variable markers ('{{', '{%', '%}', '}}'), templating will fail as
there are no matching variables.
This pull-request adds the option to configure those markers.

### CODE REVIEW

**Comment by fralau** [comment]
Hi, and thanks for your kind contribution. Could I ask you to raise an issue, to explain the problem?

**Comment by makkus** [comment]
Sure.

**Comment by angelobanse** [comment]
Hi @fralau , @makkus is right and this PR indeed fixes that issue.
:+1:

**Comment by fralau** [comment]
Re: my comment of 31 October 2018: I understand there can be a problem with variable markers, but I would really need you to open issue describing what the problem is, and how to replicate it, so that we understand what the problem is, the use case, its scope etc. Thanks.

**Comment by makkus** [comment]
Hi @fralau , I did at the time: https://github.com/fralau/mkdocs_macros_plugin/issues/5

**Comment by fralau** [comment]
As seen in #5 escaping could help in this use case:

```
{% raw %}
- task: "create a directory
  file:
    path: "{{ folder_path }}"
    state: directory
    recurse: true
{% endraw %}
```

Are there use cases that would require developing such a feature anyway?

**Comment by makkus** [comment]
I'm not using mkdocs myself anymore, but I'm dealing a bit with templating of templates. And in my experience it's easier to do that when the 'meta-template' has a different delimiter syntax. This is very specialized, and I don't imagine a lot of your users will come across it, if I were you I probably wouldn't bother with this. That being said, it's not a big change, it's just an additional optional argument to a method signature, that, when not provided, won't affect any existing code. So, up to you :-)

**Comment by DominikaK** [comment]
Here is a proposed use case:

We are interested in using this plugin in our technical documentation. However, our site contains a lot of code blocks with Twig template code (which has syntax almost identical to Jinja). The plugin interprets this code as Jinja templates and this site rendering.

It's not an option for us to rewrite all code blocks using the `{% raw %}` escaping method, because of their number. We also make use of other extensions that rely on using markdown fenced blocks.

In summary, the option proposed in this PR would actually make it possible for us to use the plugin at all.

**Comment by fralau** [comment]
Thank you for this useful information. For your use case, I understand are alreay using [Twig](https://twig.symfony.com/). Is that to produce HTML?

In essence, you wish to have _two_ stages of template processing (one with mkdocs/macros and one with Twig) with the code of both present in the same original file, is that correct?

To better understand your needs, what exact features in mkdocs/macros are the ones you need the most? [variables in the yml file, etc.?]

Do you wish to simply complement Twig with those features? Or would it be a "transition" phase, where you would like to phase Twig out?

**Comment by DominikaK** [comment]
Sorry if I was not being clear. We do not use Twig to produce HTML or markdown. The product we are documenting uses Twig, so we quote Twig code in our documentation (see for example https://github.com/ezsystems/developer-documentation/blob/master/docs/guide/templates.md#rendering-content-items-on-full-page). Because your plugin tries to interpret this code, the page cannot be rendered.

As for the plugin's functionalities, we would be mostly interested in using includes.

**Comment by fralau** [comment]
Thanks, this clarifies things a lot! I see how changing the Jinja2 syntax used by the plugin would technically avoid conflict with the Twig code in your existing documents.

On the other hand, I am not sure that I would do you a service by encouraging your to go that path. At least, I should give you my honest view about it.

Changing the syntax of a templating engine is as going out of the legal road with one's car: it seems technically easy to do at first, but after a while it will get much trickier (you can get into all sorts of weird things). And if one gets bogged or breaks something, the usual assistance services may not want to help you. So this is something one should contemplate only in case of emergency situations where the usual road will not do, and providing one is experienced with that kind of driving. Or else, if your purpose is to get a kick out of it.

If my understanding is correct, the reason you want to go that route of changing the markup syntax is not because of emergency or for kicks: it is because the usual route -- which would involve marking up the beginning and end of all your _raw code_ -- seems too long and boring.

But marking your snippets raw code as such in your source document *is the proper thing to do*.  Indeed the markdown feature that allows raw HTML/css into the text seems nice at first, but it should not be abused by adding other kinds of markup. Aside of solving this issue of preventing interpretation (which you might have with any templating engine), it would give you many advantages, e.g. easily skipping from one snippet to the other in your editor, or producing an extract containing all snippets to check that they are correct or up-to-date, etc. 

Finally, writing a little script to recognize most of your snippets, and to bracket them with the proper `raw` tags, should not be too difficult to write and could be fun. And in case you have a remorse, getting rid of those tags would be trivial.

So my best advice would be that you keep to the apparently longer but safer road, which could actually be quite rewarding.

**Comment by makkus** [comment]
As original author of this, I feel like I should chime in again: I really dont' think comparing changing template markers to 'going out of the legal road' is warranted.

Although I'm not using mkdocs itself anymore, I'm still working with templates and DSLs a lot, and I do change my template markers regularly, without any problems so far (the fact that jinja allows for this to be done easily explicitly should also be used as a data-point that this should be considered 'legal' -- just because people usually don't do it means it's bad per se). 

In a lot of those special cases we are talking about, this strategy just makes it much easier and quicker to edit templates. Using 'raw' is ok for simple use-cases, but if you work with two (logical separate) systems where the markup tokens intersect, it's much easier (and cleaner, IMHO) to make that separation clear. As long as your system is 100% upstream and separated from the underlying system (mkdocs in that case), there are not any problems at all.

I guess this is more a philosophical issue at this stage, which is fair enough I guess. It doesn't seem @fralau ever had a real need for something like this, so I can half-understand their reluctance for implementing it. From my practical experience I can guarantee that changing template markers does not introduce unforeseeable problems. I've done it a lot... From the responses to my initial pull request it seems there is *some* need for something like that, and people would like to avoid having to use 'raw' tags in some cases, for different (valid -- again IMHO) reasons... 

Also, I'd like to add, (if I remember right) my changes were very small in scope, and don't change any existing behavior. They only add options, so everyone who does not understand why other people think this is a good solution to a problem they are having can safely ignore this. Practically, I don't think this makes the code more unstable or complex in any measurable way....

**Comment by fralau** [comment]
I agree  with @makkus that the changes are minor, and I would not want to restrict the freedom of those who want to take (actually, I would go a little out of my way to give it to them).

"Philosophically", I still think that "with great power comes responsibility": changing the syntax of a templating language is "going off the beaten path" even though it is legal to do so. But certainly, it can be done safely providing one is experienced in that kind of thing.

Nevertheless, this is for experienced people. One would have to put the usual disclaimer banner that if they go that route, they are on their own and that they should not expect the developer to help them if things go wrong (much less blame the developer).

Does that seem a good approach?

**Comment by DominikaK** [comment]
Thank you, yes, it seems a good approach for us, we'd be using the plugin carefully and aware of the disclaimer.

**Comment by fralau** [comment]
I took the idea and reimplemented it (as the code had evolved since). I made a new push.
@DominikaK, let me know if it worked for you?

Follow-up in #4
